### PR TITLE
Experimenting with catching multiple errors in a test

### PR DESCRIPTION
**Pull request by scottlabs**

### CODE REVIEW

File: mocha-casperjs.js [hunk 1]
```diff
@@ -1,15 +1,37 @@
...
     }
+    function reportError() {
+        casper.checker = null
+        if (currentDone && (!currentTest || !currentTest.state)) {
+            currentDone(currentError)
+        }
+    }
+
+    if ( casper.step < casper.steps.length ) {
+        casper.run(function() {
```
Comment (nathanboktae) [inline mocha-casperjs.js]: IIRC casper doesn't work well calling `casper.run` multiple times. rather the reporting of the error should be moved down to the main `done()` call on (current) line 100.

File: mocha-casperjs.js [hunk 2]
```diff
@@ -1,15 +1,37 @@
...
     clearInterval(casper.checker)
-    casper.checker = null
-    if (currentDone && (!currentTest || !currentTest.state)) {
-      currentDone(error)
+
+    // the first error takes priority
+    if ( ! currentError ) {
+        currentError = error;
+    }
+    if ( currentTest.errors === undefined ) {
```
Comment (nathanboktae) [inline mocha-casperjs.js]: this is the mocha test object, so not sure about such a generic `errors` property. perhaps `stepErrors` ?

Comment (nathanboktae) [inline mocha-casperjs.js]: also don't `===` against undefined - rather just `!currentTest.stepErrors` or hipster oneliner `(currentTest.stepErrors || currentTest.stepErrors = []).push(error)`

**Comment by nathanboktae** [comment]
Also can you add a test case from #56  ?

**Comment by scottlabs** [comment]
Added a test case for the cascading errors and renamed to stepErrors

**Comment by scottlabs** [comment]
> IIRC casper doesn't work well calling casper.run multiple times. rather the reporting of the error should be moved down to the main done() call on (current) line 100.

I'm having trouble implementing this. As far as I understand it, by the time we're [here](https://github.com/nathanboktae/mocha-casperjs/blob/master/mocha-casperjs.js#L97) we've already got the next test and its `done` function, so any subsequent errors clobber the wrong test.

I'm down for moving calls to casper.run out of failTest but not really sure how to maintain a pointer to the current test object without calling it from failTest.

**Comment by nathanboktae** [comment]
hmm do you want to use your version in your company's test suite for a bit and let me know how it goes? I will take a crack at this in the next few days. I switched jobs so I currently don't have a real test suite to party on as of yet.

**Comment by scottlabs** [comment]
> We should not be mucking with this. I want to keep the muddling we do with casper's internal state to a minimum.

I hear ya. It's definitely a bit hacky, but otherwise waitFor errors end up hanging indefinitely.

> hmm do you want to use your version in your company's test suite for a bit and let me know how it goes? I will take a crack at this in the next few days. I switched jobs so I currently don't have a real test suite to party on as of yet.

Sounds good, I'm happy to keep experimenting. I assume there's probably a better way to implement this that doesn't rely on mucking with Casper directly but I'm not so familiar with the internals so hopefully you'll have some bright ideas :)

**Pull request merged**